### ISSUE TITLE
L059 Unnecessary quoted identifier on PostgreSQL causing case sensitivity problems

### ISSUE THREAD

**Issue by dave-v** [sqlfluff/sqlfluff#3485]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Running sqlfluff-lint and sqlfluff-fix through pre-commit, many previously quoted column names were unquoted which caused the SQL to error because the column names were now incorrect.

(These columns are pre-existing and I cannot rename them.)

### Expected Behaviour

It should not be considered unnecessary quoting if there are any uppercase characters within the quotes, as it changes the meaning of the SQL.

### Observed Behaviour

In PostgreSQL, the column name `"Column"` is not the same as `column`, but sqlfluff will consider the former a violation of L059 and rewrite it as the latter, which then causes an error when trying to execute the query.

https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

> Quoting an identifier also makes it case-sensitive, whereas unquoted names are always folded to lower case. For example, the identifiers FOO, foo, and "foo" are considered the same by PostgreSQL, but "Foo" and "FOO" are different from these three and each other. (The folding of unquoted names to lower case in PostgreSQL is incompatible with the SQL standard, which says that unquoted names should be folded to upper case. Thus, foo should be equivalent to "FOO" not "foo" according to the standard. If you want to write portable applications you are advised to always quote a particular name or never quote it.)

### How to reproduce

Here are a couple of simplified examples that illustrate the erroneous behaviour. It seems that it does quite inconsistent things depending on the surrounding context.

```sql
SELECT T."ColumnName"
FROM schema_name.table_name AS T;
```
becomes
```sql
SELECT SCHEMA_NAME.TABLE_NAME.COLUMNNAME
FROM SCHEMA_NAME.TABLE_NAME;
```
And a slightly different one:
```sql
SELECT "ColumnName"
FROM schema_name.table_name;
```
becomes
```sql
SELECT ColumnName
FROM Schema_Name.Table_Name;
```
Another:
```sql
INSERT INTO public.other_table (test_id, request_year)
SELECT "Test",
       "ReqYear"
FROM my_schema.my_table;
```
becomes
```sql
INSERT INTO public.other_table (test_id, request_year)
SELECT
    test,
    reqyear
FROM my_schema.my_table;
```

### Dialect

`postgres`

### Version

sqlfluff 1.0.0
python 3.10.2

### Configuration

`.sqlfluff`:
```
[sqlfluff]
dialect = postgres

[sqlfluff:rules]
max_line_length = 100
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
Possible short-term "fix": this rule should disable itself on Postgres (the "force_enable" behavior already present on several other rules).

For a proper fix, I wonder if we ought to add function(s) on the dialect (perhaps on the relevant segment type?) to encapsulate the appropriate identifier comparison, case sensitivity, etc. Reason: We don't want to add a lot of Postgres-specific logic to individual rules.

**Comment by tunetheweb** [comment]
I do think if your data model is using quoted column names (and IMHO best practice would be NOT to use them), then you should be manually turning this rule off anyway.

We could certainly make the default off for Postgres under the principle of avoiding breaking SQL, but feels a shame that others, not using quoted column names (which, as I say would be best practice IMHO), wouldn’t benefit from this rule then without explicitly force enabling it 😔, but it’s probably the safest thing to do despite that reservation.

**Comment by barrywhart** [comment]
Perhaps the rule could autodetect when to disable itself, based on the usage of quotes or mixed case in the file? Would need to look and think about this more carefully, but that _might_ be a good compromise between usefulness vs absolute safety.

**Comment by tunetheweb** [comment]
Or convinced it’s worth the complexity. Plus the very poor of the rule is to get consistency so not sure how you could use the existing query to figure out if the rule should be applied or not? The very case of it having “unnecessary” quotes identifiers is the case you’d want this rule to catch.

**Comment by tunetheweb** [comment]
Hey @geoHeil 

So there's a few things going on here, so please exclude the length of this reply and do let me know if it's not clear or you have any more questions.

The first is that rule L059 assumes that quotes around simple identifiers that are not reserved keywords are not needed and should be tidied up. That is not the case for certain dialects like Postgres where case-sensitive columns require these quotes.

My argument, is that it would be best practice not to use these, and if you do use them then rule L059 is not that much use to you, and should be turned off. That can be done [in a `.sqlfluff` config file](https://docs.sqlfluff.com/en/stable/configuration.html#enabling-and-disabling-rules) (using `excludes=L059`) or [inlining an exception](https://docs.sqlfluff.com/en/stable/configuration.html#enabling-and-disabling-rules) in the SQL itself if you want this rule for most of your SQL but with a few exceptions:

```sql
SELECT T."ColumnName" -- noqa: L059
FROM schema_name.table_name AS T;
```

I get the impression that some people think all rules should be followed, and excluding rules is a big step. But I think that is the wrong way to think about this, and SQLFluff, like any software, should be configured to how you want it to run. Yes the defaults should be sensible (and there are some controversial rules that perhaps should not be on by default), but even then lots of people run SQL in lots of way so we make SQLFluff configurable - including the ability to turn off rules. You should not be afraid to do this. A linter is supposed to help you, not get in your way.

So, that's one way of solving this issue for those using quoted columns in SQL - just turn the rule off! As I say best practice would not be to use these columns (as they are not as readable) in which case this rule is fine, and potentially useful to enforce a quote style, so perhaps the default _should be_ to flag them as it is now? And those that do use such columns simply turn this rule off?

However, we dislike SQLFluff breaking things in it's default set up, and clearly using `sqlfluff fix` does so here, so SQLFluff should probably be changed to handle this better. So we could do a few things to solve this
1. We could disable L059 by default and it's an explicit decision to turn it on. That would stop things breaking by default, but still would break for those that turn it on. We also currently enable all rules by default to try to encourage people to consider rules they may not have thought of, rather than have only a few rules on, and people having to realise there are more rules. Whether that's the right decision is still under some debate (see #1812) but my personal view at the moment is that it is (though I've changed my mind on this in the past!).
2. We could disable rule L059 by default for all Postgres users, and any other dialects with similar problems. We've done this in the past where rules don't really make sense for that dialect (see [rule L026 for example](https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.core.rules.Rule_L026)). We usually have a `force_enable` option then for anyone that does really want this rule for that dialect (e.g. Postgres users that do not use mixed case column names).
3. We could not apply the rule if the identifier is of mixed case? Is that the only reason to quote non-reserved words in columns? Would that solve the issue here? In that case we might want to still have a `include_mixed_case_identifers` option (off by default) to handle those. Or have that option off by default for certain dialects and on for the rest.

What's people's thoughts? I'm veering towards option 3 myself...

**Comment by geoHeil** [comment]
Understood. Thanks for the explanation -  I will disable the rule accordingly.

Regarding (1-3): I think (2) sounds the most convenient.

**Comment by tunetheweb** [comment]
Actually just noticed that this rule is already disabled for snowflake dialect for the exact same reason! OK let's go with option 2.

There is discussion in https://github.com/sqlfluff/sqlfluff/issues/3508#issuecomment-1171709762 about potentially added a completely separate rule to disallow quoted identifiers to address the concerns I had above.
